### PR TITLE
Make array conform to the Python Buffer Protocol

### LINKED ISSUE #320: Feature: Implement buffer protocol for `mlx.core` `array` to support `memoryview`
`mlx.core` arrays are great already. To support better interoperability with other frameworks, even mlx.data, implementing the [buffer protocol](https://docs.python.org/3/c-api/buffer.html) as for numpy arrays would help. 

This should address https://github.com/ml-explore/mlx-data/issues/20 and will allow to use [`memoryview`](https://docs.python.org/3/library/stdtypes.html#memoryview) from python.

### PR DESCRIPTION
**Pull request by dastrobu**
## Proposed changes

Make array conform to the [Buffer Protocol](https://docs.python.org/3/c-api/buffer.html).

The method [`__array__`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.__array__.html) is replaced by implementing the Python Buffer Protocol.

Summary:
- The complex implementation of `__array__` is removed.
- bfloat16 arrays no longer cast implicitly. For instance, `np.array(mx.ones(1, dtype=mx.bfloat16))` will fail, but `np.array(mx.ones(1, dtype=mx.bfloat16).astype(mx.float32))` will work.
- `mlx-data` may use mx arrays instead of numpy arrays, as memory views are writable. See [mlx-data issue #20](https://github.com/ml-explore/mlx-data/issues/20).
- Enables conversion to and from TensorFlow tensors without numpy, as demonstrated in `test_buffer_protocol_tf`.
- Torch may adopt the [buffer protocol in the future](https://github.com/pytorch/pytorch/issues/19143).

Closes  #320

## Writable Memory Views

Enabling writable memory views has both advantages and a drawback.

As highlighted in the summary, `mlx-data` relies on in-place modifications to buffers. If `mlx-data` were to use `mlx` arrays in the future, which is a sensible idea, `mlx` arrays must expose writable buffers, as demonstrated in this pull request. An alternative would be to refactor `mlx-data` such that it always returns new arrays on transformations and does not modify memory in place.

The downside is that direct modification of memory through buffers will not be captured by the grad tracer. Thus, this pull request allows developers to unintentionally cause issues, as illustrated in the following test case:

```py
def test_array_views_and_grad(self):
    x = mx.zeros(1)

    def f(x):
        x[:] += 1
        return x @ x

    y, df = mx.value_and_grad(f)(x)
    self.assertEqual(y.item(), 1)
    self.assertEqual(df.item(), 2)

    def g(x):
        x_view = np.array(x, copy=False)
        x_view[:] += 1 # modify memory without telling mx
        return mx.array(x @ x)

    y, dg = mx.value_and_grad(g)(x)
    self.assertEqual(y.item(), 1)
    self.assertEqual(dg.item(), 2) # this fails, as +1 is not picked up by the tracer.
```

If preventing such issues is desired, the pull request can be modified to return read-only buffers easily, albeit with the drawback of not supporting in-place operations as done in `mlx-data`. For comparison, TensorFlow has implemented the buffer protocol such that it returns read-only buffers.

From my perspective, having the flexibility to perform in-place operations on buffers outweighs the concern of potential broken gradients. I would like reviewers to pay special attention to this aspect.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: python/tests/test_array.py [hunk 1]
```diff
@@ -1039,7 +1039,7 @@ def test_memoryless_copy(self):
 
         # Check that we get read-only array that does not own the underlying data
         self.assertFalse(a_np.flags.owndata)
-        self.assertFalse(a_np.flags.writeable)
+        self.assertTrue(a_np.flags.writeable)
```
Comment (dastrobu) [inline python/tests/test_array.py:1042]: numpy seems to prefer the buffer protocol overs `__array__` internally. Since the buffer is writable, an array view is writable as well.

File: python/src/array.cpp [hunk 2]
```diff
@@ -415,6 +415,50 @@ array np_array_to_mlx(py::array np_array, std::optional<Dtype> dtype) {
   }
 }
 
+std::string buffer_format(const array& a) {
+  switch (a.dtype()) {
```
Comment (dastrobu) [inline python/src/array.cpp:419]: explicit string values determined via the unit test: 

```
self.assertEqual(mv_mx.format, mv_np.format, f"{mlx_dtype}{np_dtype}")
```
Not sure why `pybind11::format_descriptor` does not always agree with numpy.

File: python/src/array.cpp [hunk 3]
```diff
@@ -415,6 +415,50 @@ array np_array_to_mlx(py::array np_array, std::optional<Dtype> dtype) {
...
+    case int64:
+      return "l";
+    case float16:
+      return "e";
+    case float32: {
+      return pybind11::format_descriptor<float>::format();
+    }
+    case bfloat16:
+      // not numpy convertible, simply use genric "byte"
+      return "B";
```
Comment (dastrobu) [inline python/src/array.cpp:445]: use unsigned bytes as fallback, see https://docs.python.org/3/c-api/buffer.html#c.Py_buffer.format

**Comment by awni** [comment]
This looks super cool! One thing I'm wondering is what is the expected behavior when the `mx.array` goes out of scope but I still have a pointer to the memory via say `memoryview` (or another numpy array). Is that undefined? From what I can tell the array will still be freed and so will it's data so writing to the memory pointed to by the memorview is no good?

**Comment by dastrobu** [comment]
@awni thanks for the feedback. I'll be happy to contribute if this is a desired feature. 

Regarding your question: 

> Is that undefined? From what I can tell the array will still be freed and so will its data so writing to the memory pointed to by the memorview is no good?

I think this is handled by [`Py_buffer.obj`](https://docs.python.org/3/c-api/buffer.html#c.Py_buffer.obj). The python buffer holds a reference to the exporting object, which will prevent deletion of the array. So it should be well defined behavior. Let's hope pybind11 is handling it right for us. Since they [document this feature extensively](https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#buffer-protocol), I guess they will.

**Comment by dastrobu** [comment]
Even better: `__array__` could be removed altogether. There is only the edge case of auto conversion from `bfloat16` to `float32`. But it might be better to do this explicitly anyway. 
Removing `__array__` fixes https://github.com/ml-explore/mlx-data/issues/20 as well (tested locally).

@awni I updated this draft and I think it is quite a simplification and generalization compared to `__array__`. I wasn't aware of this when starting with the buffer protocol.

**Comment by awni** [comment]
This looks great to me. I don't see any obvious reasons not to go this path. It seems strictly more useful than `__array__` and keeps things pretty simple which is nice!

Adding @jagrit06 @angeloskath in case they see anything I might be missing here.

**Comment by awni** [comment]
@dastrobu I think you should PR this!

**Comment by dastrobu** [comment]
Let me check some details and add some test cases, then I'll turn the draft into a PR for review.

File: python/src/array.cpp (hunk 2 above)
Comment (dastrobu) [reply python/src/array.cpp:322]: clarified and documented via comments.

**Comment by dastrobu** [comment]
PR is ready for review. Looking forward to your feedback. 

> One thing I'm wondering is what is the expected behavior when the mx.array goes out of scope [...] Is that undefined?

@awni I added two test cases to verify references are handled as expected.

**Comment by angeloskath** [comment]
This looks great to me. I am trying to remember why we opted not to implement the buffer protocol. I can't think of any reason. However, we were returning non-writeable arrays so that it doesn't cause confusing behavior.

For instance the example provided (thanks for writing a nice document btw) is not quite right. The reason it returns 0 is because you are making a new array. If instead you were simply returning `x @ x` as in function `f`, then you would get the correct gradient. And that is way more confusing than getting 0s.

**Comment by angeloskath** [comment]
So to sum up, I like it and I will do a closer review later but I am leaning towards returning non-writeable buffers. What do you guys think @awni and @jagrit06 ?

**Comment by awni** [comment]
```
    def g(x):
        x_view = np.array(x, copy=False)
        x_view[:] += 1 # modify memory without telling mx
        return mx.array(x @ x)
```

That code is a pretty explicit situation (`copy=False` + taking grads). Dangerous words perhaps, but i can't imagine anyone expecting function transforms to "work" (not even sure what the right thing to expect there is), when you do that. 

I'm not advocating for writeable buffers, just that case doesn't have much weight for me.

What exactly do we lose by making buffers read only (the safer and hence default option)? 

- There may be occasional performance issues going to numpy. But that seems minor, and if you care about perf probably you should stay in MLX 😄 .
- Sending `mx.array` into `mx.data` ops is a nice feature. Could we get that behavior by casting to numpy in mx.data before performing the op (or something like that)?

**Comment by dastrobu** [comment]
> If instead you were simply returning x @ x as in function f, then you would get the correct gradient. And that is way more confusing than getting 0s.

> That code is a pretty explicit situation (`copy=False` + taking grads). Dangerous words perhaps, but i can't imagine anyone expecting function transforms to "work" (not even sure what the right thing to expect there is), when you do that.

Thank you for examining the example. Upon reflection, it appears to be unrelated to readonly views, as demonstrated by:

```
def g(x):
    x = np.array(x, copy=True)
    x[:] += 1
    return mx.array(x @ x)
```
This is doable with the current implementation, and of course it would also not get the right gradients. Nevertheless, unintended conversions may occur when interfacing with certain libraries. It seems advisable to highlight this behavior to developers in any case.

So  this doesn't appear to be inherently advantageous or detrimental to writable buffers.

> Sending mx.array into mx.data ops is a nice feature. Could we get that behavior by casting to numpy in mx.data before performing the op (or something like that)?

This was actually what started me to look into this. If buffers are exposed read only, mx.data should brute force write to them anyway. 

Pipelines are currently implemented such that buffers are modified in place. So if buffers should stay read only this would require some refactoring. I am not sure what performance impact that would have on pipelines operating on large datasets. 
I really like the concept of buffers and streams as they are quite powerful. But always remembering when to use a np array and when to use a mx array really causes a headache. 

Once there is agreement on the direction to go, I can update the docs accordingly.

**Comment by awni** [comment]
> This was actually what started me to look into this. If buffers are exposed read only, mx.data should brute force write to them anyway.

I thought about this a bit more, and I appreciate this use case (and similar use cases), but I don't think they are worth making the buffer writeable for. Putting MLX writeable array buffers into black box APIs seems risky in general.

Shall we go with read-only for now? Is everyone ok with that? @angeloskath @dastrobu ?

**Comment by dastrobu** [comment]
> Putting MLX writeable array buffers into black box APIs seems risky in general.

Could you elaborate on your specific concerns? Furthermore, it would be valuable to discuss the recommended approach for mlx-data or similar use cases.

Take, for instance, the `image_random_h_flip` function. It serves as an ideal model for potential extensions, with the capability to support both np and mx arrays through the buffer protocol. However, when dealing with read-only buffers, the question arises how the implementation C extension should look like?

I am open to the read-only solution, provided we can establish a clear and viable path for addressing such cases.

**Comment by awni** [comment]
@dastrobu we discussed a bit offline and the conclusion was that we can keep buffers writable (as you have it). We'll see how it goes, if there are some hiccups from we can come back and change it.

Did you have anything else to add to this PR? If not I can take a final pass and we can land it?

File: docs/src/examples/numpy.rst [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+Let's convert an array to NumPy and back.
+
+.. code-block:: python
+
+  import mlx.core as mx
+  import numpy as np
+
+  a = mx.arange(3)
+  b = np.array(a)
+  c = mx.array(b)
```
Comment (angeloskath) [inline docs/src/examples/numpy.rst:16]: This will always be a copy, perhaps it could be mentioned.

File: docs/src/examples/numpy.rst [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      return x @ x
+
+  y, df = mx.value_and_grad(f)(x)
+  print("f(x) = x² =", y.item()) # 1.0
+  print("f'(x) = 2x =", df.item()) # 2.0
+
+  def g(x):
+      x_view = np.array(x, copy=False)
+      x_view[:] += 1 # modify memory without telling mx
+      return mx.array(x @ x)
```
Comment (angeloskath) [inline docs/src/examples/numpy.rst:56]: I would remove the pure mlx example above and change this example to

```python
x_view[:] *= x_view
return x.sum()
```

with `x` being 2s or sth to clearly show that even though the output is `x**2` the gradient will be 1s.

File: docs/src/examples/numpy.rst
Comment (angeloskath) [inline docs/src/examples/numpy.rst:1]: I don't think this should be in the examples. Perhaps in `Usage`?

**Review by angeloskath — CHANGES_REQUESTED** [review]
Generally looks great! The docs need updating but other than that I think it ready to merge.

File: docs/src/examples/numpy.rst (hunk 5 above)
Comment (dastrobu) [reply docs/src/examples/numpy.rst]: Good idea. I was thinking of this example in the beginning but did not know how to express it properly.

File: docs/src/examples/numpy.rst
Comment (dastrobu) [reply docs/src/examples/numpy.rst:1]: 👍
Moved all the usages files into `usage/` to have the same folder structure as for other main sections.

File: docs/src/examples/numpy.rst (hunk 4 above)
Comment (dastrobu) [reply docs/src/examples/numpy.rst]: it is mentioned below in "By default, NumPy copies data to a new array." but I also added an in place comment for clarification.

**Comment by dastrobu** [comment]
> @dastrobu we discussed a bit offline and the conclusion was that we can keep buffers writable (as you have it). We'll see how it goes, if there are some hiccups from we can come back and change it.

Great approach. I'll be the first one to create an issue on a hiccup if I find one 😉

> Did you have anything else to add to this PR? If not I can take a final pass and we can land it?

Regarding the PR, I've incorporated all our discussions and considered the comments from @angeloskath in updating the documentation. Please take a moment to review the revised documentation. Beyond that, the PR is good to go from my perspective.

Lastly, a big thank you to all the reviewers for engaging in such a positive and constructive discussion. In my opinion, the success of open source projects is not solely dependent on good code but also on individuals like you who invest time and effort in discussions with community members like myself.

File: docs/src/usage/numpy.rst [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+representing the gradient of the sum operation alone.
+The squaring of ``x`` occurs externally to MLX, meaning that no gradient is incorporated.
+It's important to note that a similar issue arises during array conversion and copying.
+For instance, a function defined as ``mx.array(np.array(x)**2).squeeze()`` would also result in an incorrect gradient,
+even though no in-place operations on MLX memory are executed.
+
+PyTorch
+-------
+
+PyTorch supports the buffer protocol, but it requires an explicit ``memoryview``.
```
Comment (awni) [inline docs/src/usage/numpy.rst:65]: Nit `:obj:`memoryview`` will generate links.

File: docs/src/usage/numpy.rst [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  a = mx.arange(3)
+  b = torch.tensor(memoryview(a))
+  c = mx.array(b.numpy())
+
+Conversion from PyTorch tensors back to arrays must be done via intermediate NumPy arrays with ``numpy()``.
+
+TensorFlow
+----------
+
+TensorFlow supports the buffer protocol, but it requires an explicit ``memoryview``.
```
Comment (awni) [inline docs/src/usage/numpy.rst:81]: Same

File: docs/src/usage/numpy.rst [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  import mlx.core as mx
+  import torch
+
+  a = mx.arange(3)
+  b = torch.tensor(memoryview(a))
+  c = mx.array(b.numpy())
+
+Conversion from PyTorch tensors back to arrays must be done via intermediate NumPy arrays with ``numpy()``.
+
+TensorFlow
```
Comment (awni) [inline docs/src/usage/numpy.rst:78]: Could we make this a section on Jax instead of Tensorflow. I think it's more useful. Alternatively we could add a short section for Jax?

**Review by awni — APPROVED** [review]
This is so cool, thanks! I also really appreciate the nice work on the docs, that's well done! I left a couple of minor comments there could you take a look?

O/w good to go here.

File: docs/src/usage/numpy.rst (hunk 8 above)
Comment (dastrobu) [reply docs/src/usage/numpy.rst:91]: added JAX before TF.

File: docs/src/usage/numpy.rst [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+  a = mx.arange(3)
+  b = torch.tensor(memoryview(a))
+  c = mx.array(b.numpy())
+
+Conversion from PyTorch tensors back to arrays must be done via intermediate NumPy arrays with ``numpy()``.
+
+JAX
+---
+JAX fully supports the buffer protocol.
```
Comment (awni) [inline docs/src/usage/numpy.rst:80]: ❤️

**Comment by dastrobu** [comment]
> I left a couple of minor comments there could you take a look?

Thanks, everything should be resolved now.

**Review by awni — APPROVED** [review]
🚀

**Review by angeloskath — APPROVED** [review]
Looks great!
